File: hydra-haskell/src/main/haskell/Hydra/Langs/Yaml/Modules.hs
module Hydra.Langs.Yaml.Modules (moduleToYaml) where

import Hydra.Kernel
import Hydra.Adapters
import Hydra.Langs.Yaml.Serde
import Hydra.Langs.Yaml.Language
import qualified Hydra.Langs.Yaml.Model as YM
import qualified Hydra.Dsl.Types as Types

import qualified Control.Monad as CM
import qualified Data.List as L
import qualified Data.Map as M


constructModule ::
  Module Kv
  -> M.Map (Type Kv) (Coder (Graph Kv) (Graph Kv) (Term Kv) YM.Node)
  -> [(Element Kv, TypedTerm Kv)]
  -> Flow (Graph Kv) YM.Node
constructModule mod coders pairs = do
    keyvals <- withTrace "encoding terms" (CM.mapM toYaml pairs)
    return $ YM.NodeMapping $ M.fromList keyvals
  where
    toYaml (el, (TypedTerm typ term)) = withTrace ("element " ++ unName (elementName el)) $ do
      encode <- case M.lookup typ coders of
        Nothing -> fail $ "no coder found for type " ++ show typ
        Just coder -> pure $ coderEncode coder
      node <- encode term
      return (YM.NodeScalar $ YM.ScalarStr $ localNameOf $ elementName el, node)
    ns = unNamespace $ moduleNamespace mod
    localNameOf name = L.drop (1 + L.length ns) $ unName name

moduleToYaml :: Module Kv -> Flow (Graph Kv) (M.Map FilePath String)
moduleToYaml mod = withTrace ("print module " ++ (unNamespace $ moduleNamespace mod)) $ do
    node <- transformModule yamlLanguage encodeTerm constructModule mod
    return $ M.fromList [(path, hydraYamlToString node)]
  where
    path = namespaceToFilePath False (FileExtension "yaml") $ moduleNamespace mod
    encodeTerm _ = fail $ "only type definitions are expected in this mapping to YAML"

File: hydra-java/src/main/java/hydra/dsl/Modules.java
package hydra.dsl;

// import hydra.tier1.Tier1; // TODO: restore when kernel terms modules are generated
import hydra.core.Name;
import hydra.core.Term;
import hydra.core.Type;
// import hydra.graph.Element; // TODO: restore when kernel terms modules are generated
import hydra.module.Module;
import hydra.module.Namespace;
import hydra.module.QualifiedName;
import hydra.util.Opt;

// import static hydra.coreEncoding.CoreEncoding.coreEncodeType; // TODO: restore when kernel terms modules are generated


// TODO: restore when Element, Tier1, and coreEncoding are generated
public class Modules {
//    public static Element element(Module module, String localName, Term data) {
//        Name name = Tier1.unqualifyName(new QualifiedName(Opt.of(module.namespace), localName));
//        return new Element(name, data);
//    }
//
//    public static Element element(Module module, String localName, Type type) {
//        return element(module, localName, coreEncodeType(type));
//    }
//
//    public static Element element(Name name, Type type) {
//        return new Element(name, coreEncodeType(type));
//    }

    public static ModuleBuilder module(Namespace namespace) {
        return new ModuleBuilder(namespace);
    }

    public static ModuleBuilder module(Namespace namespace, String description) {
        return new ModuleBuilder(namespace, description);
    }

//    public static Name qname(Module module, String localName) {
//        return qname(module.namespace, localName);
//    }
//
//    public static Name qname(Namespace ns, String localName) {
//        return Tier1.unqualifyName(new QualifiedName(Opt.of(ns), localName));
//    }
}
